File: src/game/utils/camera.lua
local flux = require 'vendor.flux'
local Movement = require 'game.components.movement'
local Position = require 'game.components.position'
local Respawn = require 'game.components.respawn'

local Camera = {}
local LEFT_OFFSET = -WINDOW_WIDTH + 300
local RIGHT_OFFSET = -200

function Camera.new()
  local camera = {
    x = 0,
    y = 0,
    offset = RIGHT_OFFSET,
    direction = 'right',
    tween = nil
  }

  function camera:set()
    love.graphics.push()
    love.graphics.translate(
      math.floor(-self.x + 0.5),
      math.floor(-self.y + 0.5)
    )
  end

  function camera:clear()
    love.graphics.pop()
  end

  function camera:update(dt, player)
    local movement = player:as(Movement)
    local position = player:as(Position)
    local offset = self.offset

    if (player:has(Respawn)) then
      movement.direction = 'right'
    end

    if movement.direction == 'left' then
      offset = LEFT_OFFSET
    elseif movement.direction == 'right' then
      offset = RIGHT_OFFSET
    end

    if self.direction ~= movement.direction then
      self.tween = flux.group()
      self.tween:to(self, 3, {offset = offset}):ease('quadout')
    end

    if self.tween then
      self.tween:update(dt)
    end

    self.x = position.x + self.offset
    self.direction = movement.direction
  end

  return camera
end

return Camera

File: client2/src/game/utils/camera.ts
import { WINDOW_WIDTH } from '../../conf'
import { Entity } from '../../ecs/entity'
import { Movement } from '../components/Movement'
import { Position } from '../components/Position'
import * as Tween from './tween'

const LEFT_OFFSET = -WINDOW_WIDTH + 300
const RIGHT_OFFSET = -200

type Direction = 'left' | 'right'

export class Camera {
  x: number
  y: number
  offset: number
  direction: Direction
  tween?: Tween.Tween

  constructor(initialDirection: Direction) {
    this.x = 0
    this.y = 0
    this.offset = 0
    this.direction = initialDirection
  }

  clear = () => {
    love.graphics.pop()
  }

  set = () => {
    love.graphics.push()
    love.graphics.translate(
      Math.floor(-this.x + 0.5),
      Math.floor(-this.y + 0.5)
    )
  }

  update = (target: Entity) => {
    const movement = target.as(Movement)
    const position = target.as(Position)

    if (!movement || !position) {
      return
    }

    let offset = this.offset

    if (movement.direction == 'left') {
      offset = LEFT_OFFSET
    } else if (movement.direction == 'right') {
      offset = RIGHT_OFFSET
    }

    if (movement.direction !== this.direction) {
      if (this.tween) {
        this.tween.kill()
      }

      this.tween = Tween.to(
        this,
        3,
        { offset },
        {
          onUpdate: (time: number) => {
            // print(time)
          },
        }
      )
    }

    this.x = position.x + this.offset
    this.direction = movement.direction
  }
}
